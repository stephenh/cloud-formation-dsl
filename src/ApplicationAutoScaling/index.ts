
export * from './ScalableTarget';
export * from './ScalingPolicy';
