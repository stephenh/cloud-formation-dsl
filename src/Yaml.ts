
/**
 * Yaml-generation helper class.
 *
 * Currently pretty basic, should use something off-the-shelf, but this
 * is cute/easy and might end up getting prettier-type functionality.
 */
export class Yaml {

  output(top: any): string {
    let s = '';
    function concat(prefix: string, o: any) {
      Object.keys(o).forEach(k => {
        const v = o[k];
        if (v instanceof Object) {
          s += `${prefix}${k}:\n`
          concat(prefix + '  ', v);
        } else {
          s += `${prefix}${k}: ${o[k]}\n`;
        }
      });
    }
    concat('', top);
    return s;
  }
}