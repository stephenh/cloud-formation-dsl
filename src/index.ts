
/** Fn::Ref, takes any resources that has a CF name and a marker Ref attribute. */
export function Ref(o: { Name: string, Ref: string }): string {
  return `!Ref: ${o.Name}`
}

/** Shortcut for GetAttr+Arn, takes any resources that has a CF name and a marker Arn attribute. */
export function Arn(o: { Name: string, Arn: string }): string {
  return `!GetAtt: ${o.Name}.${o.Arn}`
}

/** Fn::GetAttr, takes any resources that has a CF name and a marker attribute. */
export function GetAttr<T extends { Name: string }, K extends keyof T>(o: T, k: K): string {
  return `!GetAtt: ${o.Name}.${k}`
}

/** A CloudFormation parameter. */
export function Parameter(name: string) {
  /** Parameters are ref'able, and the value is the name, which is resolved by CF. */
  return { Name: name, Ref: 'Ref' };
}