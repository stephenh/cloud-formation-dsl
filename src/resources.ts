
export const resources: [string, string, any][] = [];

