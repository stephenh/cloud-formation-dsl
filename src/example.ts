
import { AWS_ApplicationAutoScaling_ScalableTarget, AWS_ApplicationAutoScaling_ScalingPolicy } from "@src/ApplicationAutoScaling";
import { resources } from '@src/resources';
import { AWS_IAM_Role } from '@src/IAM';
import { Arn, Parameter, Ref } from '.';

const DynamoMinCapacity = Parameter("DynamoMinCapacity");
const DynamoMaxCapacity = Parameter("DynamoMaxCapacity");
const DynamoDBScaleInOutCooldown = Parameter("DynamoDBScaleInOutCooldown");
const DynamoDBTableScalingThreshold = Parameter("DynamoDBTableScalingThreshold")

const DynamoDBAutoScalingRole = AWS_IAM_Role("DynamoDBAutoScalingRole", {
  AssumeRolePolicyDocument: "..."
});

const EmployeeTableWriteCapacityScalableTarget = AWS_ApplicationAutoScaling_ScalableTarget(
  "EmployeeTableWriteCapacityScalableTarget", {
    ResourceId: "table/Employee",
    MinCapacity: Ref(DynamoMinCapacity),
    MaxCapacity: Ref(DynamoMaxCapacity),
    RoleARN: Arn(DynamoDBAutoScalingRole),
    ScalableDimension: "dynamodb:table:WriteCapacityUnits",
    ServiceNamespace: "dynamodb"
  });

const EmployeeTableWriteScalingPolicy = AWS_ApplicationAutoScaling_ScalingPolicy(
  "EmployeeTableWriteScalingPolicy", {
    PolicyName: "EmployeeTableWriteScalingPolicy",
    PolicyType: "TargetTrackingScaling",
    ScalingTargetId: Ref(EmployeeTableWriteCapacityScalableTarget),
    TargetTrackingScalingPolicyConfiguration: {
      TargetValue: Ref(DynamoDBTableScalingThreshold),
      ScaleInCooldown: Ref(DynamoDBScaleInOutCooldown),
      ScaleOutCooldown: Ref(DynamoDBScaleInOutCooldown),
      PredefinedMetricSpecification: {
        PredefinedMetricType: "DynamoDBWriteCapacityUtilization"
      }
    }
  });

const EmployeeTableReadCapacityScalableTarget = AWS_ApplicationAutoScaling_ScalableTarget(
  "EmployeeTableReadCapacityScalableTarget", {
    ResourceId: "table/Employee",
    MinCapacity: Ref(DynamoMinCapacity),
    MaxCapacity: Ref(DynamoMaxCapacity),
    RoleARN: Arn(DynamoDBAutoScalingRole),
    ScalableDimension: "dynamodb:table:ReadCapacityUnits",
    ServiceNamespace: "dynamodb"
  });

const EmployeeTableReadScalingPolicy = AWS_ApplicationAutoScaling_ScalingPolicy(
  "EmployeeTableReadScalingPolicy", {
    PolicyName: "EmployeeTableReadScalingPolicy",
    PolicyType: "TargetTrackingScaling",
    ScalingTargetId: Ref(EmployeeTableReadCapacityScalableTarget),
    TargetTrackingScalingPolicyConfiguration: {
      TargetValue: Ref(DynamoDBTableScalingThreshold),
      ScaleInCooldown: Ref(DynamoDBScaleInOutCooldown),
      ScaleOutCooldown: Ref(DynamoDBScaleInOutCooldown),
      PredefinedMetricSpecification: {
        PredefinedMetricType: "DynamoDBReadCapacityUtilization"
      }
    }
  });

resources.forEach(resource => {
  const [name, type, props] = resource
  console.log(name)
})
