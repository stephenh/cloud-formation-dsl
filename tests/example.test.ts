
import { resources } from '../src/resources';
import '../src/example';
import { Yaml } from '../src/Yaml';

it('should have a few resources', () => {
  expect(resources.length).toBe(5);
  const y = new Yaml();
  const output = {};
  resources.forEach(resource => {
    const [name, type, props] = resource
    output[name] = {
      Type: type,
      Properties: props
    }
  })
  expect(y.output(output)).toMatchSnapshot();
});