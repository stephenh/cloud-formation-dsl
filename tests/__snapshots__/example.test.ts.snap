// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have a few resources 1`] = `
"DynamoDBAutoScalingRole:
  Type: AWS::IAM::Role
  Properties:
    AssumeRolePolicyDocument: ...
EmployeeTableWriteCapacityScalableTarget:
  Type: AWS::ApplicationAutoScaling::ScalableTarget
  Properties:
    ResourceId: table/Employee
    MinCapacity: !Ref: DynamoMinCapacity
    MaxCapacity: !Ref: DynamoMaxCapacity
    RoleARN: !GetAtt: DynamoDBAutoScalingRole.Arn
    ScalableDimension: dynamodb:table:WriteCapacityUnits
    ServiceNamespace: dynamodb
EmployeeTableWriteScalingPolicy:
  Type: AWS::ApplicationAutoScaling::ScalingPolicy
  Properties:
    PolicyName: EmployeeTableWriteScalingPolicy
    PolicyType: TargetTrackingScaling
    ScalingTargetId: !Ref: EmployeeTableWriteCapacityScalableTarget
    TargetTrackingScalingPolicyConfiguration:
      TargetValue: !Ref: DynamoDBTableScalingThreshold
      ScaleInCooldown: !Ref: DynamoDBScaleInOutCooldown
      ScaleOutCooldown: !Ref: DynamoDBScaleInOutCooldown
      PredefinedMetricSpecification:
        PredefinedMetricType: DynamoDBWriteCapacityUtilization
EmployeeTableReadCapacityScalableTarget:
  Type: AWS::ApplicationAutoScaling::ScalableTarget
  Properties:
    ResourceId: table/Employee
    MinCapacity: !Ref: DynamoMinCapacity
    MaxCapacity: !Ref: DynamoMaxCapacity
    RoleARN: !GetAtt: DynamoDBAutoScalingRole.Arn
    ScalableDimension: dynamodb:table:ReadCapacityUnits
    ServiceNamespace: dynamodb
EmployeeTableReadScalingPolicy:
  Type: AWS::ApplicationAutoScaling::ScalingPolicy
  Properties:
    PolicyName: EmployeeTableReadScalingPolicy
    PolicyType: TargetTrackingScaling
    ScalingTargetId: !Ref: EmployeeTableReadCapacityScalableTarget
    TargetTrackingScalingPolicyConfiguration:
      TargetValue: !Ref: DynamoDBTableScalingThreshold
      ScaleInCooldown: !Ref: DynamoDBScaleInOutCooldown
      ScaleOutCooldown: !Ref: DynamoDBScaleInOutCooldown
      PredefinedMetricSpecification:
        PredefinedMetricType: DynamoDBReadCapacityUtilization
"
`;
